airbyte-cdk/python/airbyte_cdk/test/http/mocker.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+import contextlib
+import functools
+from types import TracebackType
+from typing import Callable, List, Optional, Union
+
+import requests_mock
+from airbyte_cdk.test.http import HttpRequest, HttpRequestMatcher, HttpResponse
+
+
+class HttpMocker(contextlib.ContextDecorator):
+    """
+    WARNING: This implementation only works if the lib used to perform HTTP requests is `requests`
+    """
+
+    def __init__(self) -> None:
+        self._mocker = requests_mock.Mocker()
+        self._matchers: List[HttpRequestMatcher] = []
+
+    def __enter__(self) -> "HttpMocker":
+        self._mocker.__enter__()
+        return self
+
+    def __exit__(self, exc_type: Optional[BaseException], exc_val: Optional[BaseException], exc_tb: Optional[TracebackType]) -> None:
+        self._mocker.__exit__(exc_type, exc_val, exc_tb)
+
+    def _validate_all_matchers_called(self) -> None:
+        for matcher in self._matchers:
+            if not matcher.has_expected_match_count():
+                raise ValueError(f"Invalid number of matches for `{matcher}`")
+
+    def get(self, request: HttpRequest, responses: Union[HttpResponse, List[HttpResponse]]) -> None:
+        """
+        WARNING: Given multiple requests that are not mutually exclusive, the request will match the first one. This can happen in scenarios
+        where the same request is added twice (in which case there will always be an exception because we will never match the second
+        request) or in a case like this:
+        ```
+        http_mocker.get(HttpRequest(_A_URL, headers={"less_granular": "1", "more_granular": "2"}), <...>)
+        http_mocker.get(HttpRequest(_A_URL, headers={"less_granular": "1"}), <...>)
+        requests.get(_A_URL, headers={"less_granular": "1", "more_granular": "2"})
+        ```
+        In the example above, the matcher would match the second mock as requests_mock iterate over the matcher in reverse order (see
+        https://github.com/jamielennox/requests-mock/blob/c06f124a33f56e9f03840518e19669ba41b93202/requests_mock/adapter.py#L246) even
+        though the request sent is a better match for the first `http_mocker.get`.
+        """
+        if isinstance(responses, HttpResponse):
```

Comment:
nit: do we really get a lot of benefit from allowing the developer to pass in a single response or a list of responses? It might just be more straightforward to drop the Union and just say pass in a list of responses. given the variable name is already called `responses`. I mean this is trivial but just a thought

Reply:
I think there is as if there are only one response, the fact that it's a list is not important and the `[]` will clutter the tests. This is even more important as having only one response for a request is the default behavior. 

If I were in Java, we would have:
```
def get(self, request: HttpRequest, response: HttpResponse) -> None:
    self.get(request, [response])
    
def get(self, request: HttpRequest, responses: List[HttpResponse]) -> None:
    # do something
```

I don't know how to emulate this in Python without the Union. I feel like having another method name is not really relevant here because that's the same behavior we want from both calls. Let me know what you prefer based on this reasonning!

Reply:
mmk i think its fine to leave it as you have it. this was just a nit and if it makes the test writing experience a little easier then 👍